[LETS-154] Improve performance of parallel recovery/replication

cristiarg opened the pull request:
http://jira.cubrid.org/browse/LETS-154

Improve performance of parallel recovery/replication.

Patch is made up of two main changes:
1.  refactoring:
    - the functionality provided by `minimum_log_lsa_monitor` has been integrated into member class `redo_parallel::min_unapplied_log_lsa_monitoring`
    - `redo_job_queue` became superfluous and has also been integrated partly in `redo_parallel` and partly in `redo_parallel::redo_task`
    - class `redo_parallel::redo_task` has expanded in scope to also handle internally the consumption of jobs (previously handled by `redo_job_queue`)
2. implementation change:
    - async tasks are currently created and managed by `redo_parallel`
    - a pushed job is distributed upfront to a certain async task based on its `VPID` with help from a hashing function (`hash<VPID>`)
    - `redo_task` will manage internally the jobs' consumption as needed using the approach of two swapped containers, one on the "push" (produce) side - `m_produce_vec`, and one on the "pop" (consume) side - `redo_task::execute` - `jobs_vec`
    - after each job has been executed, it is `retire`d at which point its ownership goes back to the reusable job container
    - updated unit-tests to test changes:
      - some of the tests have been deleted because the corresponding tested entities became internal implementation details
      - the core unit-test functor `execute_test` has been adapted to fully cover `redo_parallell` surface
    - `min_unapplied_log_lsa_monitoring` performs calculation pro-actively via an internal thread by interrogating the async tasks but will also be triggered from the `wait_past_target_log_lsa` function when the latter is called

Other:
- extended reported time spend in `log_recovery_redo` with the time needed for "setting up" (this includes pre-populating reusable job container with desired number of jobs, allocating memory for internals and starting up infrastructure)
- added a `log_lsa::is_max` function
- added a `hash<VPID>` implementation based on examples in Boost libraries

cristiarg committed:
transform adress into pointer to be able to supply nullptr when in recovery mode

cristiarg committed:
transform 'raw' queues of jobs in maps of vpid and 'raw' queues; other optimizations

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-157-maps-of-vpids

 Conflicts:
	src/transaction/log_recovery.c
	src/transaction/log_recovery_redo_parallel.cpp
	src/transaction/log_recovery_redo_parallel.hpp

cristiarg committed:
use intermediary std::map<log_lsa, vpid> to always consume the vpid with the minimum log lsa such as to avoid the overall minimum log lsa having to jump back and forth, thereby misleading the replication algorithm which relies on it

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-157-maps-of-vpids

cristiarg committed:
remove helper/debug code

cristiarg committed:
remove helper/debug code

cristiarg committed:
streamline consume erase from maps

cristiarg committed:
code cleanup

cristiarg committed:
rename function

cristiarg committed:
remove repeated use of 'cubthread::get_entry', transform to member variable

cristiarg committed:
add timing for parallel recovery

cristiarg committed:
refactor towards creating an affinity between vpid and thread index by means of a hash<VPID>; WiP; no minimum lsa monitoring code yet

cristiarg committed:
cs access mode

cristiarg committed:
proper 'cubthread::get_entry' usage via getter iso ctor

cristiarg committed:
Revert "cs access mode"

This reverts commit 2a8d9260f9ae96c0fbb6dd2f45a8a74d2870e84e.

cristiarg committed:
expose critical section access mode as a parameter for the functions that perform log page reading

cristiarg committed:
log reader uses 'LOG_CS_SAFE_READER' iso 'LOG_CS_FORCE_USE'

cristiarg committed:
don't monitor min lsa

cristiarg committed:
set main thread type as 'TT_MASTER'

cristiarg committed:
add perf stats to measure where time is spent in the recovery main thread

cristiarg committed:
thread type recovery identifies under a user module iso of as a system module

cristiarg committed:
comment out perf for max time value and for the per-transaction all values

cristiarg committed:
replace perf time tracking with a nanosecond-level metric based on std::chrono

cristiarg committed:
only start perf recording at the start of the log recovery redo sequence

cristiarg committed:
prompt when to actually start log recovery sequence (used to know when to actually launch 'perf record' for the cub_server process)

cristiarg committed:
inline all log reading logic: log_reader and the legacy functions used by log_reader

cristiarg committed:
cubrid.conf used for performance testing of log recovery

cristiarg committed:
larger config values for a scaled up test

cristiarg committed:
re-enable prompt to know when to trigger perf profiling

cristiarg committed:
more discrete perf stats for redo or push async part of the recovery redo

cristiarg committed:
proper perf when waiting for idle

cristiarg committed:
make std::hash a member

cristiarg committed:
comment out perf logging in log_recovery_redo function

cristiarg committed:
add perf logging to evaluate time spent in redo_job_queue::push_job

cristiarg committed:
memoization for vpid -> task_index mapping via hash function

cristiarg committed:
avoid re-creating per-task job container upon each attempted pop; replace deque with vector and reserve a number of values upon creation

cristiarg committed:
deactivate perf

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

 Conflicts:
	src/transaction/log_recovery.c
	src/transaction/log_recovery_redo_parallel.hpp
	src/transaction/log_replication.cpp

cristiarg committed:
merge corrections

cristiarg committed:
remove locale

cristiarg committed:
update config for extended statistics flag

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

cristiarg committed:
clean-up perf constants

cristiarg committed:
move away from unique_ptr in preparation of adding a new intermediate layer between main thread and worker threads based on lockfree_circular_queue

cristiarg committed:
introduce a lockfree queue processed by a thread between the main thread and the log redo applying threads in order to relieve the main thread from the time spent locking and distributing jobs

cristiarg committed:
minor labelling correction

cristiarg committed:
larger lockfree queue buffer

cristiarg committed:
new perf logging for new-ing a job to be dispatched

cristiarg committed:
counter and timer

cristiarg committed:
introduce job implementation without templates in preparation for reuse of job instances

cristiarg committed:
temporary commented out unit test

cristiarg committed:
introduce a stack of reusable jobs to function as an allocator

cristiarg committed:
measure perf for pre-allocation stack

cristiarg committed:
reuse internally pre-reserved vector of pointers used to exchange jobs between queue and recovery redo applying threads

cristiarg committed:
introduce back the version where the lockfree circular queue is not used; guarded by a pre-processor variable

cristiarg committed:
steady_clock for perf instead of high_resolution_clock

cristiarg committed:
log number of reusable jobs in stack

cristiarg committed:
better tuning constants

cristiarg committed:
measure time spent setting up parallel recovery redo infrastructure (with all pre-allocations)

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

Conflicts:
  src/transaction/log_recovery_redo_parallel.cpp
  src/transaction/log_recovery_redo_parallel.hpp

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

 Conflicts:
	src/transaction/log_reader.cpp
	src/transaction/log_reader.hpp
	unit_tests/log/test_main_data_page_fetcher.cpp
	unit_tests/log/test_main_log_page_fetcher.cpp

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

 Conflicts:
	src/transaction/log_reader.hpp

cristiarg committed:
remove unused lockfree queue usage

cristiarg committed:
optimize usage of 'logtb_find_tran_index'

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

Conflicts:
  src/transaction/log_recovery.c
  src/transaction/log_recovery_redo_parallel.cpp
  src/transaction/log_recovery_redo_parallel.hpp
  src/transaction/log_recovery_redo_perf.hpp
  src/transaction/log_replication.cpp
  unit_tests/log/test_main_log_recovery_parallel.cpp

cristiarg committed:
measure blocking_pop

cristiarg committed:
unsigned -> std::size_t

cristiarg committed:
try to reduce contention between main thread and recovery redo applying threads on the reverse path that returns job pointers back to main thread for reuse

cristiarg committed:
try to reduce contention between main thread and recovery redo applying threads on the reverse path that returns job pointers back to main thread for reuse

cristiarg committed:
remove perf entry

cristiarg committed:
extract constant

cristiarg committed:
log flush back count

cristiarg committed:
alignment

cristiarg committed:
measure perf overhead

cristiarg committed:
vector is deque

cristiarg committed:
allocate jobs in-place in pre-allocated contiguous memory

cristiarg committed:
remove redundant perf stat logging

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

Conflicts:
  src/transaction/log_recovery.c

cristiarg committed:
spacing

cristiarg committed:
add param recovery_reusable_jobs_count

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

 Conflicts:
	src/transaction/log_reader.hpp
	src/transaction/log_recovery.c
	src/transaction/log_recovery_redo_parallel.cpp
	src/transaction/log_recovery_redo_parallel.hpp
	src/transaction/log_replication.cpp

cristiarg committed:
comments

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

 Conflicts:
	src/transaction/log_recovery.c
	src/transaction/log_recovery_redo_parallel.cpp
	src/transaction/log_recovery_redo_parallel.hpp
	src/transaction/log_recovery_redo_perf.hpp

cristiarg committed:
pass pointer by reference to surface its value in the calling function

cristiarg committed:
incorrect condition for single not initialization

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

cristiarg committed:
guard when perfmon activation flags are not used (the default) to log performance statistics for async reconvery threads

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity

Conflicts:
  src/transaction/log_recovery.c
  src/transaction/log_recovery_redo_parallel.cpp
  src/transaction/log_recovery_redo_parallel.hpp
  src/transaction/log_replication.cpp

cristiarg committed:
fix unit tests by commenting out some inimplemented/to-be-disabled stuff related to 'wait-for-idle'

cristiarg committed:
refactor out redo_job_queue:
- remove 'wait_for_idle' and 'is_idle'
- in replicator, replace 'wait_for_idle' with 'wait_for_termination_and_stop_execution' because the lifecycle of the replication is clearer now than it was when first implemented; the code is untested though as it can only be tested when 'minimum_unreplicated_log_lsa' functionality is re-implemented

cristiarg committed:
refactor out red_job_queue:
- rename function

cristiarg committed:
refactor out red_job_queue:
- remove commented out and dead code from redo_job_queue

cristiarg committed:
refactor out redo_job_queue:
- remove job queue as a middle layer and integrate tasks and dispatch to tasks directly into redo_parallel

cristiarg committed:
refactor out redo_job_queue:
- remove redo_job_queue stale code

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity-with-refactor

cristiarg committed:
implemented minimum not applied log_lsa for recovery; updated unit tests

cristiarg committed:
deactivate minimum not-applied log_lsa for recovery

cristiarg committed:
more & proper asserts

cristiarg committed:
comparison

cristiarg committed:
reset changes in cubrid.conf

cristiarg committed:
delete accidentally added conf file

cristiarg committed:
reset perf monitor changes

cristiarg committed:
reset ts_timer changes

cristiarg committed:
remove overhead perf entry

cristiarg committed:
strict comparison

cristiarg committed:
remove instrumentation code; streamline the code that measures time taken by setting_up, main and async parts

cristiarg committed:
do not forward hardcoded constant as a parameter

cristiarg committed:
remove system parameter which was needed for testing

cristiarg committed:
swap unique_ptr's instead of raw pointers

cristiarg committed:
beefed up 'wait_past_target_lsa' part; removed redundant old unit tests and unit tests that could no longer be performed because of refactoring (ie: minimum_log_lsa)

cristiarg committed:
set start redo log_lsa upfront for replication in order to avoid race conditions to occur at a possible call to wait_past_target_lsa

cristiarg committed:
introduce call to function set_outer_not_applied_log_lsa in unit tests

cristiarg commented on src/transaction/log_recovery_redo_parallel.hpp:
```diff
@@ -311,9 +156,30 @@ namespace cublog
+       * synchronously by the main thread and the calculation of the minimum not-applied log_lsa needs
+       * to be accurate
+       */
+      std::atomic<log_lsa> m_outer_not_applied_log_lsa;
```
The `_not_applied_log_lsa` members and associated functions can be abstracted away in an implementation detail class.

ribram replied:
Do you think the class may be nested into redo_parallel?

ribram replied:
Can "outer" be replaced with something more suggestive? "main thread" maybe? Or something that we can use to reference that main execution thread everywhere?

cristiarg replied:
Done

cristiarg commented on src/transaction/log_recovery_redo_parallel.hpp:
```diff
@@ -120,7 +65,7 @@ namespace cublog
     public:
```
at least class `redo_job_base` should be moved out of `redo_parallel`;
maybe also `redo_task`

cristiarg commented on src/transaction/log_replication.cpp:
```diff
@@ -173,11 +170,12 @@ namespace cublog
   void
   replicator::conclude_task_execution ()
```
to be removed

ribram commented on src/transaction/log_lsa.hpp:
```diff
@@ -104,12 +105,19 @@ log_lsa::log_lsa (const int64_t value)
+  return *this == MAX_LSA;
+}
+
+
```
```suggestion

```

ribram commented on src/transaction/log_recovery.c:
```diff
@@ -1014,15 +1016,19 @@ log_recovery_redo (THREAD_ENTRY * thread_p, log_recovery_context & context)
+  const auto time_dur_setting_up_ms = std::chrono::duration_cast <std::chrono::milliseconds>(
+        std::chrono::system_clock::now () - time_start_setting_up);
+  // *INDENT-ON*
+  const auto time_start_main = std::chrono::system_clock::now ();
```
- A single `std::chrono::system_clock::now ()` call, under disabled indentation
- Avoid breaking after parenthesis if possible
- Nitpicking: remove `time_` from `time_dur_` to shorten the name.
```suggestion
  // *INDENT-OFF*
  const auto time_start_main = std::chrono::system_clock::now ();
  const auto dur_setting_up_ms =
    std::chrono::duration_cast <std::chrono::milliseconds> (time_start_main - time_start_setting_up);
  // *INDENT-ON*
```

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -500,15 +99,67 @@ namespace cublog
 
       log_rv_redo_context m_redo_context;
+
+      std::unique_ptr<redo_job_vector_t> m_produce_vec;
```
Why unique pointer to vector?
```suggestion
      redo_job_vector_t m_produce_vec;
```

cristiarg replied:
Evolution: raw pointer -> unique pointer -> vector.

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
   redo_parallel::redo_task::execute (context_type &context)
   {
-    bool finished = false;
+    std::unique_ptr<redo_job_vector_t> jobs_vec { new redo_job_vector_t () };
```
Why unique pointer to vector?
```suggestion
    redo_job_vector_t jobs_vec;
```

cristiarg replied:
Done

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+  }
+
+  inline void
+  redo_parallel::redo_task::pop_jobs (std::unique_ptr<redo_parallel::redo_task::redo_job_vector_t> &a_out_job_vec)
```
```suggestion
  redo_parallel::redo_task::pop_jobs (redo_job_vector_t &a_out_job_vec)
```

cristiarg replied:
Done

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
-	if (job == nullptr && adding_finished)
+	if (!jobs_popped && m_adding_finished.load ())
 	  {
 	    finished = true;
```
In situations like these, it is usual to use `while (true)` loop and break it when it's finished.
```suggestion
	    break;
```

cristiarg replied:
Done

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+    assert (jobs_vec->empty ());
+
     m_task_state_bookkeeping.set_inactive ();
+    m_in_execution.store (false);
```
`m_in_execution` seems to do the same thing as `m_task_state_bookkeeping`. Is it necessary?

cristiarg replied:
Replaced with an expanded task state bookkeeping which also helps get rid of `m_in_execution`.

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+      std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
+      is_vector_empty = m_produce_vec->empty ();
+    }
+    return is_vector_empty && !m_in_execution.load ();
```
Too complicated code for a simple check.
```suggestion
    std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
    return m_produce_vec->empty () && !m_in_execution.load ();
```

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+	// still empty, which is then found and tested-only by the pop side
+	assert (expected_log_lsa <= a_log_lsa);
+      }
+    while (!m_not_applied_log_lsa.compare_exchange_strong (expected_log_lsa, a_log_lsa));
```
There is no race here to need `compare_exchange_strong`. There is only one situation when main thread changes `m_not_applied_log_lsa`, when both queues are empty and calling `set_not_applied_log_lsa_from_pop_side` is not possible.

The worker thread may simply set the new value:
```suggestion
    assert (m_do_monitor_not_applied_log_lsa);
    assert (m_not_applied_log_lsa <= a_log_lsa);

    m_not_applied_log_lsa = a_log_lsa;
```

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+  inline void
+  redo_parallel::redo_task::set_not_applied_log_lsa_from_push_side (const log_lsa &a_log_lsa)
+  {
+    assert (m_do_monitor_not_applied_log_lsa);
+
+    {
+      const log_lsa snapshot_not_applied_log_lsa { m_not_applied_log_lsa.load () };
+      // strict comparison because jobs have ever-increasing log_lsa's
+      assert (snapshot_not_applied_log_lsa == MAX_LSA || snapshot_not_applied_log_lsa < a_log_lsa);
+    }
+
+    log_lsa expected_not_applied_log_lsa { MAX_LSA };
+    m_not_applied_log_lsa.compare_exchange_strong (expected_not_applied_log_lsa, a_log_lsa);
+    // TODO: any check for the return value?
```
`m_not_applied_log_lsa = MAX_LSA` is set only with the protection of m_produce_vec_mtx and once set, it may not be changed back. The `compare_exchange_strong` may not be necessary.
```suggestion
    assert (m_not_applied_log_lsa == MAX_LSA || m_not_applied_log_lsa < a_log_lsa);
    if (m_not_applied_log_lsa == MAX_LSA)
      {
        m_not_applied_log_lsa = MAX_LSA;
      }
```

cristiarg replied:
Removed.

ribram reviewed (commented):
Only partially reviewed; I will continue to review later.

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+    if (m_do_monitor_not_applied_log_lsa && m_produce_vec->empty ())
+      {
+	const log_lsa &job_log_lsa = a_job->get_log_lsa ();
+	set_not_applied_log_lsa_from_push_side (job_log_lsa);
```
Is there a reason to prefer an extra variable and two-line code, instead of no variable and one line code?
```suggestion
	set_not_applied_log_lsa_from_push_side (a_job->get_log_lsa ());
```

cristiarg replied:
Done

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -736,6 +523,107 @@ namespace cublog
+  {
+    assert (m_do_monitor_minimum_log_lsa);
+
+    log_lsa minimum_not_applied_log_lsa { m_outer_not_applied_log_lsa.load () };
```
Add a comment explaining that the calculus has to start with the LSA of main thread (outer).

ribram replied:
Add assert is is not MAX_LSA.
```suggestion
    log_lsa minimum_not_applied_log_lsa { m_outer_not_applied_log_lsa.load () };
    assert (minimum_not_applied_log_lsa < MAX_LSA);
```

cristiarg replied:
Added an initial `log_lsa` as a ctor parameter and asserted prerequisite in the ctor.

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -736,6 +523,107 @@ namespace cublog
+      std::scoped_lock<std::mutex> slock { m_calculate_minimum_not_applied_log_lsa_mtx };
+      calculated_minimum_not_applied_log_lsa = m_calculated_minimum_not_applied_log_lsa;
+    }
+
```
No lock is required to do the early out.
```suggestion
    if (a_target_lsa < m_calculated_minimum_not_applied_log_lsa)
      {
        // Already past
        return;
      }
```

cristiarg replied:
Done

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity-with-refactor

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+  inline void
+  redo_parallel::redo_task::push_job (redo_parallel::redo_job_base *a_job)
+  {
+    std::scoped_lock<std::mutex> slock { m_produce_vec_mtx };
```
This is my first encounter with scoped_lock so I did a bit of reading to understand the differences between various locks. I tend to agree with the conclusions of this [response](https://stackoverflow.com/a/60172828):

> So my advice is to use the simplest tool for the job:
>  1. lock_guard if you need to lock exactly 1 mutex for an entire scope.
>  2. scoped_lock if you need to lock a number of mutexes that is not exactly 1.
>  3. unique_lock if you need to unlock within the scope of the block (which includes use with a condition_variable).

cristiarg replied:
Replaced

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+		      {
+			set_not_applied_log_lsa_from_pop_side (job->get_log_lsa ());
+		      }
+		    job->execute (thread_entry, m_redo_context);
```
Is the time separation between execution and retirement irrelevant?

cristiarg replied:
Fixed

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -517,135 +168,287 @@ namespace cublog
+
+    const vpid &job_vpid = a_job->get_vpid ();
+    const std::size_t vpid_hash = m_vpid_hash (job_vpid);
+    const std::size_t task_index = vpid_hash % m_task_count;
```
The successions of operations is easier to be read in a single line:
```suggestion
    const std::size_t task_index = m_vpid_hash (a_job->get_vpid ()) % m_task_count;
```

cristiarg replied:
Done

ribram commented on src/transaction/log_recovery_redo_parallel.hpp:
```diff
@@ -311,9 +156,30 @@ namespace cublog
+      /* following members control the pro-active calculation of the minimum not-applied log_lsa as well
+       * as means to actually trigger and wait for the calculation asynchronously
+       */
+      log_lsa m_calculated_minimum_not_applied_log_lsa;
```
To reduce the name size, `minimum_not_applied_log_lsa` may be replaced by `min_unapplied_lsa`.

cristiarg replied:
Done

ribram commented on unit_tests/log/test_main_log_recovery_parallel.cpp:
```diff
@@ -64,43 +67,104 @@ void execute_test (const log_recovery_test_config &a_test_config,
 	{
+	  if (a_test_config.wait_past_previous_log_lsa)
+	    {
+	      log_redo_parallel.set_outer_not_applied_log_lsa (job->get_log_lsa ());
```
Normally `set_outer_not_applied_log_lsa()` should be called for each log record, regardless if it is applied by the main therad or it will be applied by a worker thread.

cristiarg replied:
Done.

cristiarg committed:
review remarks

cristiarg committed:
- avoid race condition in wait_for_termination_and_stop_execution due to the way that function termination was signalled
- introduce a conditional wait in the pop_jobs function to avoid having to sleep in the task execute function
- for the above, add a new function to notify a task to terminate waiting in case it is still waiting for a job when the termination condition is met
- introduce more checks when setting the not_applied_log_lsa value and avoid the atomic test & set routines

cristiarg committed:
make the test more relevant; set the outer minimum log_lsa for each applied job

cristiarg committed:
no need for an unique pointer for vector

cristiarg committed:
separate perf logging for execute and retire

cristiarg committed:
replace internal flag variable with expanded active state bookkeeping class

cristiarg committed:
in unit tests, do not wait past last added job; when setting outer log_lsa, immediately notify such that the calculating thread proactively calculates the new value

cristiarg committed:
Revert "in unit tests, do not wait past last added job; when setting outer log_lsa, immediately notify such that the calculating thread proactively calculates the new value"

This reverts commit 2907539fc5e2ff6ad4bda2cb4dcb9dbd5fbdb90e.

cristiarg committed:
- in unit tests, do not wait past last added job;
- when setting outer log_lsa, immediately notify such that the calculating thread proactively calculates the new value
- a start log_lsa is required by redo_parallel ctor for consistency
- minor review remarks

cristiarg committed:
renames

cristiarg committed:
calculated min unapplied log_lsa can never be max

ribram commented on src/transaction/log_recovery.c:
```diff
@@ -997,13 +997,15 @@ log_recovery_redo (THREAD_ENTRY * thread_p, log_recovery_context & context)
 
   volatile TRANID tran_id;
   volatile LOG_RECTYPE log_rtype;
+  const auto time_start_setting_up = std::chrono::system_clock::now ();
```
```suggestion
  // *INDENT-OFF*
  const auto time_start_setting_up = std::chrono::system_clock::now ();
  // *INDENT-ON*
```

cristiarg replied:
really needed? indent does not seem to be broken in anyway

ribram replied:
Right now it is not broken, but it becomes unpredictable. It may break after other changes, with no apparent relations.

cristiarg committed:
retire function is not needed anymore

cristiarg committed:
refactor min unapplied log_lsa in a nested class

cristiarg committed:
shorter names

cristiarg committed:
variable can be approximated by other conditions

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -24,652 +24,620 @@
+  redo_parallel::redo_task::is_idle () const
+  {
+    std::lock_guard<std::mutex> lockg { m_produce_vec_mtx };
+    return m_produce_vec.empty () && !m_task_state_bookkeeping.is_active (m_task_idx);
```
Doesn't `!m_task_state_bookkeeping.is_active (m_task_idx)` imply `m_produce_vec.empty ()`?

cristiarg replied:
it does

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -24,652 +24,620 @@
-	return nullptr;
+	set_unapplied_log_lsa_from_push (a_job->get_log_lsa (), lockg);
+
+	m_produce_vec_cv.notify_one ();
```
Notifying before unlocking `m_produce_vec_mtx` may wake-up thread only to get blocked on the mutex. Unlock the mutex before the notification.

cristiarg replied:
done

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -24,652 +24,620 @@
-	// constantly swapping two empty containers; works in combination with a notification
-	// dispatched after new work items are added to the produce queue; this, in effect means that
-	// this function will semantically become 'blocking_pop'
+    m_produce_vec_cv.notify_one ();
```
I think a mutex lock/unlock is required before calling `m_produce_vec_cv.notify_one ()`. Otherwise this is possible:

- Worker calls `pop_jobs`, loads m_adding_finished false.
- Main thread sets adding finished to true immediately after and tries to notify the worker
- The workers fails to pass condition and starts waiting on the condition variable right after missing the notification.

By forcing a mutex lock/unlock, the worker can no longer miss the notification; it is either already waiting on the condition variable or it will check m_adding_finished after it is modified.

cristiarg replied:
done

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -24,652 +24,620 @@
+	m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_POP);
+	if (jobs_vec.empty () && m_adding_finished.load ())
+	  {
+	    break;
```
```suggestion
	if (m_adding_finished.load ())
	  {
	    assert (jobs_vec.empty ());
	    break;
```

cristiarg replied:
The target is to consume all available jobs, not to finish when signaled.
Even if the flag is set to true from the main thread, a task may still have one more batch of jobs to process. See the wait lambda in `pop_jobs` which has two exit paths:
- either jobs are still to be processed (in which case the value of the flag is irrelevant, and might be true which would invalidate the proposed new assert)
- or there are no jobs, in which case the flag becomes the stop condition

ribram replied:
Right, disregard my suggestion.

cristiarg committed:
avoid a rarely occuring dead-lock:
- task active state monitoring must notify state change to inactive for each task because monitoring is nominal now
- when pushing a job, the internal task notification must be fired out of the lock

cristiarg committed:
Merge remote-tracking branch 'upstream/scalability_dev' into lets-154-vpid-affinity-with-refactor

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -736,6 +705,26 @@ namespace cublog
+  redo_parallel::get_calculated_min_unapplied_log_lsa ()
+  {
+    return m_min_unapplied_log_lsa_calculation.get_calculated_log_lsa ();
+  }
```
`redo_parallel::get_calculated_min_unapplied_log_lsa ()` and `min_unapplied_log_lsa_calculation::get_calculated_log_lsa ()` are only used in unit test; they don't seem to be useful for the engine. My suggestion is to remove them.

cristiarg replied:
removed

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -24,652 +24,617 @@
+	  std::unique_lock<std::mutex> ulock { m_calculate_mtx };
+	  // wait might be interrupted by an outside notify and this is expected
+	  m_calculate_cv.wait_for (ulock, std::chrono::milliseconds (1000));
+	}
```
Wait time of 1 second is too large for a system that has only one active transaction; if notification occurs right before waiting, the transaction will be blocked for one second. Reduce the wait time to 10 msec.

cristiarg replied:
noticed that too; done already

ribram replied:
it does not look done, the wait time is still 1000 msec.

cristiarg replied:
Sorry. My mistake.

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -24,652 +24,617 @@
+    std::unique_lock<std::mutex> ulock { m_calculate_mtx };
+    m_calculate_cv.wait (ulock, [this, &a_target_lsa] ()
+    {
+      //assert (!m_adding_finished.load ());
 
```
```suggestion
```

cristiarg replied:
done

ribram commented on src/transaction/log_recovery_redo_parallel.cpp:
```diff
@@ -24,652 +24,617 @@
+    for (auto &redo_task: m_redo_tasks)
+      {
+	redo_task->notify_adding_finished ();
+      }
```
Since every task has to be notified, an atomic shared m_adding_finished might not be useful; maybe each task should get its own local m_adding_finished that gets set when notify_adding_finished is called.

cristiarg replied:
Done. And it does not need to be atomic.

ribram commented on src/transaction/log_recovery_redo_parallel.hpp:
```diff
@@ -135,170 +79,124 @@ namespace cublog
        */
       void wait_for_termination_and_stop_execution ();
 
+      inline bool get_waited_for_termination () const
+      {
+	// all steps from the wait_for_termination.. blocking function must have been done
+	return m_adding_finished.load ()
+	       && !m_task_state_bookkeeping.is_any_active ()
+	       && (m_worker_pool == nullptr);
+      }
```
This function is incomprehensible; there is no such thing as being waited for termination :P. The code may be moved directly where the check is needed.

cristiarg replied:
removed, asserts were already part of the dtor

ribram commented on src/transaction/log_recovery_redo_parallel.hpp:
```diff
@@ -135,170 +79,124 @@ namespace cublog
-	  /* decrement the internal number of active tasks
-	   */
-	  inline void set_inactive ();
+	  const std::vector<std::unique_ptr<redo_task>> *m_redo_tasks;
```
```suggestion
	  const std::vector<std::unique_ptr<redo_task>> &m_redo_tasks;
```

cristiarg replied:
Done. But having it as a pointer allowed to spot a race condition which might have slipped otherwise.

cristiarg committed:
avoid a rarely occuring dead-lock:
- force a lock right before notifying a task that adding has finished
- in min unapplied log_lsa dtor, avoid race condition upon thread termination where redo tasks were cleared before the calculating thread has joined

cristiarg committed:
remove incomprehensible function' already part of the dtor

cristiarg committed:
pass by reference instead of pass by pointer

cristiarg committed:
one adding finished flag for each task

cristiarg committed:
adding finished flag does not need to be atomic

cristiarg committed:
remove function used only in uni tests

cristiarg committed:
reduce wait time

ribram reviewed (approved):
Please fix https://github.com/CUBRID/cubrid/pull/3000#discussion_r704313688 and merge.
